File: Facebook Hacker Cup/Round 1/a2.cpp

#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <math.h>
using namespace std;

std::ifstream fin("a2real.txt");
ofstream fout("a2out.txt");

int t;

// bool validate(int interval, vector<int> cows, int n, int m , int c) {
//     if( (x < n) || (buses > m)) return false;
//     return true;
// }

bool isStartX(vector<char> w) {
    auto firstX = find(w.begin(), w.end(), 'X');
    auto firstO = find(w.begin(), w.end(), 'O');
    bool onX;
    if (firstO == w.end() || (firstX - w.begin() < firstO - w.begin() && firstX != w.end())) {
        onX = true;
    }
    else {
        onX = false;
    }
    return onX;
}

long long solve(int n, vector<char> w) {
    bool onX = isStartX(w);
    bool prevStartX = onX;

    long long cnt = 0;
    long long total = 0;

    vector<char> xo;
    vector<int> xoind;

    for (int ind = 0; ind < w.size(); ++ind) {

        if (w[ind] == 'X'|| w[ind] == 'O') {
            xo.push_back(w[ind]);
            xoind.push_back(ind);
        }

        if (onX && w[ind] == 'O') {
            cnt++;
            onX = false;
        }
        else if (!onX && w[ind] == 'X') {
            cnt++;
            onX = true;
        }
        total += cnt;
        //cout << cnt << "\n";
    }

    if (xo.size() == 0) {
        return 0;
    }
    //cout << "First total: " << total << "\n";

    //cout << prevStartX << " " << (xo.front() == 'X') << "\n";
    bool currStartX;
    long long result = 0;
    for (int i = 0; i < n-1; i++) {
        currStartX = (xo.front() == 'X');
        if (currStartX != prevStartX) {
            prevStartX = currStartX;
            total -= n - xoind.front();
        }
        if (total <= 0) {
            break;
        }
        //cout << total << " ";
        result += total;
        
        if (w.front() == xo.front()) {
            xo.erase(xo.begin());
            xoind.erase(xoind.begin());
        }
        w.erase(w.begin());
    }
    //cout << "\n";
    return result % 1000000007;
}

int main() {
    fin >> t;

    for (int a=0; a<t; a++) {
        int n;
        vector<char> w;
        fin >> n;
        char c;
        for (int b=0; b<n; b++){
            fin >> c;
            w.push_back(c);
        }
        long long solution = solve(n, w);

        fout << "Case #" << a+1 << ": " << solution << "\n";
        //cout << solve(n, w) << " <-- result" << "\n";
        //cout << "Case #" << a+1 << ": " << solution << "\n";
    }
    
}

File: Facebook Hacker Cup/Round 1/a2.py
import collections
import random


# test = open("a1tester.txt", 'w')
# test.write("80\n")
# letters = ["F", "O", "X"]
# for b in range(80):
#     test.write("50000\n")
#     for a in range(50000):
#         l = letters[random.randint(0, 2)]
#         test.write(l)
#     test.write("\n")



fin = open("a2tester.txt", 'r')
fout = open("a2out.txt", 'w')


t = int(fin.readline().strip())


def solve():
    changeList = []
    n = int(fin.readline().strip())
    w = fin.readline().strip()
    switch = 0
    firstO, firstX = w.find("O"), w.find("X")
    if firstO == -1 or (firstX < firstO and firstX != -1):
        onX = True
        if firstX == -1:
            return 0
        changeList.append(firstX)
    else:
        onX = False
        if firstO == -1:
            return 0
        changeList.append(firstO)


    for c in range(len(w)):

        if w[c] == "X":
            if not onX:
                switch += 1
                onX = True
                changeList.append(c) 
        elif w[c] == "O":
            if onX:
                switch += 1
                onX = False
                changeList.append(c)
    
    return switch
for case in range(t):
    fout.write("Case #{}: {}\n".format( case + 1, solve() ))

fout.close()
